File: javajava/PalindromoRecursivo.java

public class PalindromoRecursivo {
    public static void main(String[] args) {
      processInput();
      MyIO.print("FIM");
    }

    public static void processInput() {
        String palavra = MyIO.readLine();
        int i=0;
        if (!Comparar(palavra, i)){
            if(IsPalindromo(palavra, i)==true){
                MyIO.println("SIM");
            }
            else{
                MyIO.println("NAO");
            } //Método para chamar permitir que a main "se chame" dentro dela mesma.
            processInput();
        } 
    }

    public static boolean Comparar(String palavra, int i){
        String palavra2="FIM";
        if(i<palavra.length()){
            if(i>=palavra.length() || palavra.charAt(i) != palavra2.charAt(i)){
                return false;
            } //Compara o tamanho e letra por letra das duas palavras para ver se uma delas e igual a FIM.
            i++;
            return Comparar(palavra,i);
        }
        return true;
    }


    static boolean IsPalindromo(String palavra, int i){
        if(i<palavra.length()){
        if(palavra.charAt(i)==palavra.charAt(palavra.length()-1-i)){
            i++;
            return IsPalindromo(palavra,i);
        }
        else{
            return false;
        }
        }//Checa letra por letra para ver se a palavra é um palíndromo
      return true;
    }
}

File: javajava/PalindromoRecursivo.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

bool Comparar(const char *palavra, int i);
bool IsPalindromo(const char *palavra, int i);


int main() {
    processInput();
    printf("FIM\n");
    return 0;
}


void processInput() {
    char palavra[100];
    if (fgets(palavra, sizeof(palavra), stdin) != NULL) {
        int i = 0;
   
        if (!Comparar(palavra, i)) {
            if (IsPalindromo(palavra, i)) {
                printf("SIM\n");
            } else {
                printf("NAO\n");
            }
            processInput();
        }
    }
}

bool Comparar(const char *palavra, int i) {
    const char *palavra2 = "FIM";
    if (i < strlen(palavra)) {
        if (i >= strlen(palavra) || palavra[i] != palavra2[i]) {
            return false;
        }
        i++;
        return Comparar(palavra, i);
    }//Compara se a palavra é igual a FIM, letra por letra e pelo tamanho
    return true;
}

bool IsPalindromo(const char *palavra, int i) {
    if (i < strlen(palavra)) {
        if (palavra[i] == palavra[strlen(palavra) - 1 - i]) {
            i++;
            return IsPalindromo(palavra, i);
        } else {
            return false;
        }
    }//Checa se uma palavra é um palíndromo ao checar letra por letra.
    return true;
}

